src/gax/src/paginator.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use futures::stream::unfold;
+use futures::{Stream, StreamExt};
+use pin_project::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+
+/// Describes a type that can be iterated over asyncly when used with [Paginator].
+pub trait PageableResponse {
+    fn next_page_token(&self) -> String;
+}
+
+/// An adapter that converts list RPCs as defined by [AIP-4233](https://google.aip.dev/client-libraries/4233)
+/// into a [futures::Stream] that can be iterated over in an async fashion.
+#[pin_project]
+pub struct Paginator<T, E> {
+    #[pin]
+    stream: Pin<Box<dyn Stream<Item = Result<T, E>>>>,
+}
+
+impl<T, E> Paginator<T, E>
+where
+    T: PageableResponse,
+{
+    /// Creates a new [Paginator] given the initial page token and a function
+    /// to fetch the next [PageableResponse].
+    pub fn new<F>(seed_token: String, execute: impl Fn(String) -> F + Clone + 'static) -> Self
+    where
+        // TODO(codyoss): This, and other types, may also need send here in practice.
+        F: Future<Output = Result<T, E>> + 'static,
+    {
+        let stream = unfold(IterState::Token(seed_token), move |state| {
+            let execute = execute.clone();
```

Comment:
Why do we need to clone this? Feels like we should be able to borrow the function to call it, that is why it is not a `FnOnce`, right?

Reply:
I believe this is needed because Fn does not implement copy and we move this this function  to the async block below for the call. There may be a better way to call this, but this is where I ended up after a well fought battle with the compiler. If you see away around this I am open, just could not figure out anything better.